File: 백준/Silver/1417. 국회의원 선거/국회의원 선거.py
import heapq

N = int(input())
ds_vote_cnt = int(input())
etc_vote_cnt_heap = []
res = 0


def check_top_rank():
    # 다솜이가 제일 득표수가 많을 때 끝낸다
    if ds_vote_cnt > -etc_vote_cnt_heap[0]:
        return True
    else:
        return False


for _ in range(1, N):
    num = int(input())
    heapq.heappush(etc_vote_cnt_heap, -num)  # 최대힙

if len(etc_vote_cnt_heap) == 0:
    print(res)
else:
    while True:
        if check_top_rank():
            break

        # etc에서 최대값 가져오기
        etc_max_val = -heapq.heappop(etc_vote_cnt_heap)

        # 다솜 +1, 최대값 -1
        ds_vote_cnt += 1
        etc_max_val -= 1

        # 최대값 다시 힙에 넣기
        if etc_max_val > 0:
            heapq.heappush(etc_vote_cnt_heap, -etc_max_val)

        res += 1

    print(res)

File: 백준/Silver/1417. 국회의원 선거/국회의원 선거.cc
#include <iostream>
#include <queue>

using namespace std;

bool check_top_rank(int ds_vote_cnt, int etc_max_val) {
    return ds_vote_cnt > etc_max_val;
}

int main() {
    int N, ds_vote_cnt, num, res = 0;
    priority_queue<int> etc_vote_cnt_heap;

    cin >> N >> ds_vote_cnt;

    for (int i = 1; i < N; ++i) {
        cin >> num;
        etc_vote_cnt_heap.push(num);
    }

    if (etc_vote_cnt_heap.empty()) {
        cout << res << endl;
        return 0;
    }

    while (true) {
        if (check_top_rank(ds_vote_cnt, etc_vote_cnt_heap.top())) {
            break;
        }

        int etc_max_val = etc_vote_cnt_heap.top();
        etc_vote_cnt_heap.pop();

        ds_vote_cnt += 1;
        etc_max_val -= 1;

        if (etc_max_val > 0) {
            etc_vote_cnt_heap.push(etc_max_val);
        }

        res += 1;
    }

    cout << res << endl;
    return 0;
}
